### PR TITLE
allow constructor names in substrate

### PR DESCRIPTION
**Pull request by xermicus**
This changes a few things regarding constructor names for the substrate target:

- Constructors may now have a name, like functions.
- If they have no name, `new` will be used automatically.
  - It is a convention to call "constructors" `new` in Rust too
  - `new` is a reserved keyword in Solidity, so you can't write a function called `new` anyways (prevents conflicts)
- It works nicely together with the function name mangling.

To make instantiation of `ink!` (Substrate) contracts possible in the future, I think  generating interface files is the better solution than changing our syntax to allow it.

### CODE REVIEW

File: src/sema/contracts.rs [hunk 1]
```diff
@@ -793,6 +800,28 @@ fn substrate_requires_public_functions(contract_no: usize, ns: &mut ast::Namespa
...
+/// Constructors and functions are no different in Substrate.
+/// This function checks that all constructors and function names are unique.
+/// Overloading (mangled function or constructor names) is taken into account.
+fn substrate_unique_constructor_names(contract_no: usize, ns: &mut ast::Namespace) {
+    if !ns.target.is_substrate() || ns.diagnostics.any_errors() {
+        return;
+    }
+
+    let mut functions = HashMap::new();
+    for f in ns.contracts[contract_no].functions.iter() {
```
Comment (seanyoung) [inline src/sema/contracts.rs:812]: Very pedantic, but is `.iter()` really necessary here.

File: docs/language/contracts.rst [hunk 2]
```diff
@@ -18,18 +18,21 @@ like so:
       }
   }
 
-A constructor does not have a name and may have any number of arguments. If a constructor has arguments,
-then when the contract is deployed then those arguments must be supplied.
+A constructor can have any number of arguments.
+If a constructor has arguments, those arguments must be supplied when the contract is deployed.
 
 If a contract is expected to hold receive value on instantiation, the constructor should be declared ``payable``.
```
Comment (seanyoung) [inline docs/language/contracts.rst:24]: Please fix `hold receive` to `receive` (or something like that)

File: docs/language/contracts.rst [hunk 3]
```diff
@@ -18,18 +18,21 @@ like so:
...
 
 .. note::
+    In Substrate, constructors have a name. Solang allows to name constructors in the substrate target:
```
Comment (seanyoung) [inline docs/language/contracts.rst:27]: Not so sure about `In Substrate`. The documentation says `On Substrate` already a lot. We should at least be consistent.

File: docs/language/contracts.rst [hunk 4]
```diff
@@ -18,18 +18,21 @@ like so:
...
-  Solana. So, when building for Substrate, multiple constructors can be
-  defined as long as their argument list is different (i.e. overloaded).
+    .. code-block:: solidity
+ 
+        contract Foo {
+            constructor my_new_foo() {}
+        }
 
-  When the contract is deployed in the Polkadot UI, the user can select the constructor to be used.
+    Unnamed constructors will be called ``new`` in the metadata.
```
Comment (seanyoung) [inline docs/language/contracts.rst:35]: I think we should mention that the constructor name is only used for the metadata and cannot be used for with the `new contract()` syntax.

**Review by seanyoung — CHANGES_REQUESTED** [review]
Just some minor documentation comments. Looks good!

File: docs/language/contracts.rst [hunk 5]
```diff
@@ -18,18 +18,24 @@ like so:
       }
   }
 
-A constructor does not have a name and may have any number of arguments. If a constructor has arguments,
-then when the contract is deployed then those arguments must be supplied.
+A constructor can have any number of arguments.
+If a constructor has arguments, those arguments must be supplied when the contract is deployed.
```
Comment (LucasSte) [inline docs/language/contracts.rst:22]:
```suggestion
If a constructor has arguments, they must be supplied when the contract is deployed.
```
This is less wordy.

File: docs/targets/substrate.rst [hunk 6]
```diff
@@ -9,7 +9,7 @@ The Parity Substrate has the following differences to Ethereum Solidity:
 - The address type is 32 bytes, not 20 bytes. This is what Substrate calls an "account"
 - An address literal has to be specified using the ``address"5GBWmgdFAMqm8ZgAHGobqDqX6tjLxJhv53ygjNtaaAn3sjeZ"`` syntax
 - ABI encoding and decoding is done using the `SCALE <https://docs.substrate.io/reference/scale-codec/>`_ encoding
-- Multiple constructors are allowed, and can be overloaded
+- Constructors can be named. Constructors with no name will be called "new" in the generated metadata.
```
Comment (LucasSte) [inline docs/targets/substrate.rst:12]:
```suggestion
- Constructors can be named. Constructors with no name will be called ``new`` in the generated metadata.
```
This is to maintain consistency with your past usage of `new`.

**Review by LucasSte — CHANGES_REQUESTED** [review]

**Review by seanyoung — APPROVED** [review]

File: docs/language/contracts.rst [hunk 7]
```diff
@@ -18,18 +18,24 @@ like so:
...
-A constructor does not have a name and may have any number of arguments. If a constructor has arguments,
-then when the contract is deployed then those arguments must be supplied.
+A constructor can have any number of arguments.
+If a constructor has arguments, they must be supplied when the contract is deployed.
 
-If a contract is expected to hold receive value on instantiation, the constructor should be declared ``payable``.
+If a contract is expected to receive value on instantiation, the constructor should be declared ``payable``.
 
 .. note::
+    On Substrate, constructors have a name. Solang allows to name constructors in the substrate target:
```
Comment (LucasSte) [inline docs/language/contracts.rst:27]:
```suggestion
    On Substrate, constructors have a name. Solang allows naming constructors in the substrate target:
```
See: https://english.stackexchange.com/questions/60271/grammatical-complements-for-allow/60285#60285

**Review by LucasSte — APPROVED** [review]
Just one nitpick.

**Pull request merged**